### DISCUSSION TITLE
How to define a schema containing a property belonging to a custom type

### DISCUSSION THREAD

**Discussion by leungkinghin-ct** [sinclairzx81/typebox#1054] [Q&A]
I am new to typebox and have a question. Currently in our application we defined a schema as following

```
export const responseSchema = Type.Object({
  resultCode: Type.String(),
  userId: Type.String(),
});
```

Now we would like to make a change on this schema so it also supports the entire custom type `User` instead of only `userId` as following
```
export type User = { 
   userId: string
   version: number
   createdBy: string
   ... 
   ...
   ...


}
```

```
export const responseSchema = Type.Object({
  resultCode: Type.String(),
  user: Type.Object(User), // which introduce compilation error
});
```

In this case, how can we define the `responseSchema` by using `User` directly instead of extracting every properties from `User` in schema definition?

Thanks

**Comment by sinclairzx81** [comment]
@leungkinghin-ct Hi,

You can use generic types for this. The following creates a generic ResponseSchema that includes the `resultCode` and allows for additional properties to be specified by generic parameter.

[TypeScript Link Here](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBoUAUoSazAYDOOAyjAIYzADGcAvnAGYEhwDkAAqcADt6AG2rAoAehjoMAIwgAPLgChlEgFTq4AcQwCMUBihkoIcAMJQMtDHABKZSANK2qUAK70Y7q3HUTleghneAdSJxcKegALDBBqOABeOAAefEIDOjI4DAUYPQATUjwCIizSAD4ACjBSzJJSAC44dLKGgEokiuNMADoAeVkAKwwvKqRlOCm4K1J3YRhzCAKMZrQ+t0EAcyr2nEnp3qPajOJs5SZ21QkJOABBZoBVFyhlaUx7pLgqWgYU94wEFY9wqgWCpHgdy+YQiGCisXi4wOU3cLzWMgGw1GMCR0zxcFRBgAkgV0RsYIYBDs9sj8QA3Az8YJkjC9ABy7hAsgMuywtLx9CsNgKACFUCzepsqbz+dMbnAjr1aZcLlc1LcRU8XiS3iYRV8fnR6P8ZEC4CLQUEQuboY5wXCYnFqLjpoSoCSJVLqaqgA)

```typescript
import { Type, TProperties, Static } from '@sinclair/typebox'

/** Generic Type To Create Response Structure */
const ResponseSchema = <Properties extends TProperties>(properties: Properties) => Type.Object({ 
  resultCode: Type.String(), 
  ...properties 
})

// A: User
type A = Static<typeof A>
const A = ResponseSchema({
  user: Type.Object({
    userId: Type.String(),
    version: Type.Number(),
    createdBy: Type.String(),
    // ...
  })
})

// B: UserId
type B = Static<typeof B>
const B = ResponseSchema({
  userId: Type.String()
})
```

Hope this helps
S

**Reply by leungkinghin-ct** [reply]
@sinclairzx81 Thanks for your example and support. 

I am afraid I haven't explained clearly. My main challenge here is to avoid spreading the properties from `User` type. It is because `User` type in my project is defined in another lib which is a huge type contains nested object and a lot of properties. I hope it can be reused in my typebox schema but not re-define once again as following approach
```
export const responseSchema = Type.Object({
  resultCode: Type.String(),
  user: Type.Object({
       userId: Type.String(),
       version: Type.Number(),
       createdBy: Type.String(),
    // ... too many properties and objects below so I am too lazy to redefine again here

   }), 
});
```

That's why I would like to know if it is possible define a schema with a custom type directly without spreading its all properties. Thanks.

**Reply by sinclairzx81** [reply]
@leungkinghin-ct Heya,

> That's why I would like to know if it is possible define a schema with a custom type directly without spreading its all properties. Thanks.

I'm not sure I follow, but going over your original post, you had the following TypeScript type.. This is defined as follows.

```typescript
export type User = { 
   userId: string
   version: number
   createdBy: string
   ... 
   ...
   ...
}
```
Then you suggested that the following raises a compilation error.

```typescript
export const responseSchema = Type.Object({
  resultCode: Type.String(),
  user: Type.Object(User), // which introduce compilation error
});
```
Not sure if the `User` type is a typo, but this error is expected. TypeBox types cannot be constructed from raw TS static types. They need to be constructed as TypeBox types in order to be composed. This is because TypeScript types are erased at compilation. What this means is you will need to replicate the remote library types as TypeBox types in your project.

```typescript
export const User = Type.Object({
   userId: Type.String(),
   version: Type.Number(),
   createdBy: Type.String()
   ... 
   ...
   ...
})

export const responseSchema = Type.Object({
  resultCode: Type.String(),
  user: User, // no error
});
```
There tooling available to help you convert TS types to TypeBox types, you can find this at the link below.

[TypeBox Workbench Link Here](https://sinclairzx81.github.io/typebox-workbench/?share=ZXhwb3J0IHR5cGUgVXNlciA9IHsgCiAgIHVzZXJJZDogc3RyaW5nCiAgIHZlcnNpb246IG51bWJlcgogICBjcmVhdGVkQnk6IHN0cmluZwogICAvLyAuLi5jb3B5IGFuZCBwYXN0ZSBuZXN0ZWQgcHJvcGVydGllcyBoZXJlCn0KCmV4cG9ydCB0eXBlIHJlc3BvbnNlU2NoZW1hID0gewogIHJlc3VsdENvZGU6IHN0cmluZwogIHVzZXI6IFVzZXIKfQo%3D)

Alternatively, if you don't care about the schematics, you can just do the following to create a custom type via Unsafe (this isn't recommended, but will let you embed the remote library static type in TypeBox schematics)

```typescript
type User = { ... }

export const responseSchema = Type.Object({
  resultCode: Type.String(),
  user: Type.Unsafe<User>(Type.Any()) // will give correct inference, but will not validate
});

```

Does this help?
